47degrees/sbt-energymonitor
# Add demo persistence server

This PR adds an implementation of a server that will be compatible with the shape of energy testing output that #17 will introduce.

It consists of:

- two compose files, one for bringing up the database on its own, and one for running migrations against it
- some enhanced dependencies to make testing easier
- accidentally introducing more effectful testing libs (didn't realize I already had Weaver in the plugin 🤦🏻 so now there's also munit + scalacheck + cats-effect in the application code)
- cross-publication of new resources for Scala.js for secret future plans (it would probably be nice to run a server that's only used intermittently and in focused bursts in a Lambda context, rather than keeping another VM on 24/7, and I'm convinced by the [arguments in Feral's readme](https://github.com/typelevel/feral/tree/v0.1.0-M8#why-go-feral))

Testing
----

(all example requests use httpie)

- `./scripts/update` to pull / build containers and run migrations
- `docker-compose up -d` to bring up the server
- create a measurement:

```
echo '{"joules": 23.0, "seconds": 8, "organization": "47degrees", "repository": "sbt-energymonitor", "branch": "abcde", "run": 8, "recordedAt": "2022-05-05T10:15:00Z"}' | http :8080
```

- list the measurements for `47degrees/sbt-energymonitor`: `http :8080/47degrees/sbt-energymonitor`
- list the measurements for `47degrees/sbt-energymonitor` on a branch that doesn't exist -- `http :8080/47degrees/sbt-energymonitor branch==some-bogus-branch`
- use the `to-run` and `from-run` query params to find runs in `47degrees/sbt-energymonitor` with run numbers in the `[from-run, to-run]` range (e.g., from 7 to 9 should return the one you posted above, from 10 (no to specified) should not, to 7 (no from specified) should not, etc.)
- check the mean energy consumption on this branch (there's only one entry so far so it's easy to tell what it should be): `http :8080/47degrees/sbt-energymonitor/mean/abcde` (you can also filter the mean by the from / to params for the list endpoint if you want but it's not very interesting with one entry)
